File: src/main/java/com/zerozero/core/infrastructure/rabbitmq/MessageProducer.java
package com.zerozero.core.infrastructure.rabbitmq;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.amqp.rabbit.core.RabbitTemplate;

@Log4j2
@RequiredArgsConstructor
public abstract class MessageProducer<T extends BaseQueueProperty, R> {

    private final RabbitTemplate rabbitTemplate;

    protected final T queueProperty;

    public void publishMessage(R request) {
        try {
            rabbitTemplate.convertAndSend(queueProperty.getExchange(), queueProperty.getRoutingKey(), request);
        } catch (Exception e) {
            log.error("[MessageProducer] - publishMessage Failed", e);
        }
    }
}

File: src/main/java/com/zerozero/core/infrastructure/rabbitmq/MessageProducer.kt
package com.zerozero.core.infrastructure.rabbitmq

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate

abstract class MessageProducer<T : BaseQueueProperty, R : Any>(
    private val rabbitTemplate: RabbitTemplate,
    private val queueProperty: T
) {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    fun publishMessage(request: R) {
        try {
            rabbitTemplate.convertAndSend(queueProperty.exchange, queueProperty.routingKey, request)
        } catch (e: Exception) {
            log.error("[MessageProducer] - publishMessage Failed", e)
        }
    }
}
